src/utils/cloud.ts

```diff
@@ -107,6 +107,302 @@ export const getDeviceAndMaybeAppFromUUID = _.memoize(
+	const cacheStream = await getImageWritableStream(
+		options.deviceType,
+		options.version,
+	);
+
+	pass.pipe(cacheStream, { end: false });
+	pass.on('end', cacheStream.persistCache);
+
+	// If we return `pass` directly, the client will not be able
+	// to read all data from it after a delay, since it will be
+	// instantly piped to `cacheStream`.
+	// The solution is to create yet another PassThrough stream,
+	// pipe to it and return the new stream instead.
+	const pass2 = new PassThrough() as InstanceType<typeof PassThrough> & {
+		mime: string;
+	};
+	pass2.mime = imageStream.mime;
+	imageStream.on('progress', (state) => pass2.emit('progress', state));
+
+	imageStream.on('error', async (err) => {
+		await cacheStream.removeCache();
+		pass2.emit('error', err);
+	});
+
+	return pass.pipe(pass2);
+};
+
+/**
+ * @summary Get a device operating system image
+ * @function
+ * @public
+ *
+ * @description
+ * This function saves a copy of the downloaded image in the cache directory setting specified in [balena-settings-client](https://github.com/balena-io-modules/balena-settings-client).
+ *
+ * @param {String} deviceType - device type slug or alias
+ * @param {String} versionOrRange - can be one of
+ * * the exact version number,
+ * in which case it is used if the version is supported,
+ * or the promise is rejected,
+ * * a [semver](https://www.npmjs.com/package/semver)-compatible
+ * range specification, in which case the most recent satisfying version is used
+ * if it exists, or the promise is rejected,
+ * * `'latest'` in which case the most recent version is used, including pre-releases,
+ * * `'recommended'` in which case the recommended version is used, i.e. the most
+ * recent version excluding pre-releases, the promise is rejected
+ * if only pre-release versions are available,
+ * * `'default'` in which case the recommended version is used if available,
+ * or `latest` is used otherwise.
+ * Defaults to `'latest'`.
+ * @param {Object} options
+ * @param {boolean} options?.developmentMode
+ * @returns {Promise<NodeJS.ReadableStream>} image readable stream
+ *
+ * @example
+ * manager.get('raspberry-pi', 'default').then (stream) ->
+ * 	stream.pipe(fs.createWriteStream('foo/bar.img'))
+ */
+const getStream = async (
+	deviceType: string,
+	versionOrRange: string,
+	options: Omit<DownloadConfig, 'deviceType' | 'version'> = {},
+) => {
+	if (versionOrRange == null) {
+		versionOrRange = 'latest';
+	}
+	const version = await resolveVersion(deviceType, versionOrRange);
+	const isFresh = await isImageFresh(deviceType, version);
+	const $stream = isFresh
+		? await getImage(deviceType, version)
+		: await doDownload({ ...options, deviceType, version });
+	// schedule the 'version' event for the next iteration of the event loop
+	// so that callers have a chance of adding an event handler
+	setImmediate(() =>
+		$stream.emit('balena-image-manager:resolved-version', version),
+	);
+	return $stream;
+};
+
```

Comment:
Can we move this into a separated file (it can be under utils) like `image-manager.ts` or something along this lines? Otherwise this `cloud.ts` modules will just grow forever